### DISCUSSION TITLE
Mutate same session cookie in different loaders

### DISCUSSION THREAD

**Discussion by miniplus** [remix-run/remix#3900] [Q&A]
In the case of a parent/child route structure, on initial page load, both loaders are called.
In the following example we use the same session (cookie based) in both of these loaders.

Both loaders have the request as a starting point and use that to get the cookie based session.
Obviously, we now have 2 instances of that cookie session which we are mutating.

This causes 2 problems:

1. By committing the session headers, you are only committing that 1 specific mutation, not both.
2. Because it's an initial page load and both loaders are called, we are setting the same Set-Cookie header twice with a different value, causing a race condition (?)

How can I manage the same session across multiple loaders and not set the cookie multiple times?

```typescript
const { getSession, commitSession } =
  createCookieSessionStorage({Path: '/'});

// /parent
export const loader = async ({ request }) => {
    const session = await getSession(request.headers.get('Cookie'));
    session.set('parent', 'parentValue');

    return json({}, {
        headers: new Headers({
            'Set-Cookie': await commitSession(session)
        })
    })
}

// /parent/child
export const loader = async ({ request }) => {
    const session = await getSession(request.headers.get('Cookie'));
    session.set('child', 'childValue');

    return json({}, {
        headers: new Headers({
            'Set-Cookie': await commitSession(session)
        })
    })
}
```

**Comment by kiliman** [comment]
You should typically update session on actions, which are guaranteed to only be one at any given time. Why are you setting session in your loader?

If we can understand your usecase better, we may be able to come up with a solution.

**Reply by miniplus** [reply]
The main reason is calling an external API on page load.
If this call fails, my thought was to `session.flash` a message onto the session, as we can then use that flash message in the component to display the failed API call to the user.

In case of a parent/child route structure you need to `session.flash` and commit the session in both loaders.

> You should typically update session on actions

My use case seems to be a completely valid reason to update the session, unless I'm thinking along a completely wrong architectural path.

**Reply by kiliman** [reply]
Yeah, I mentioned in your SO post that perhaps cookie backed sessions are not the best way to handle this. The problem is that with cookie sessions, the shared state is also stored independently in the cookie (per request). So when you have multiple requests running in parallel, each has an independent view of the session state.

**Reply by kiliman** [reply]
Another option is to simply return the error message from your loader
```js
function loader() {
  const result = await callApi()
  if (result.error) {
    return json({error: result.error })
  }
  return json({data: result.data})
}
export default function() {
  const { data, error } = useLoaderData()
  return (
    { error && <Toast message={error}/> }
    ...
  )
}
```

**Reply by kiliman** [reply]
P.S. Session flash is good if you need to pass a value/message across redirects. But if you're on the same route, it's typically better to just return that data in the loader response.

Also, if you need to access the results from the parent loader, you can `useMatches` to get the loader data from any active route.

**Reply by miniplus** [reply]
>Yeah, I mentioned in your SO post that perhaps cookie backed sessions are not the best way to handle this. The problem is that with cookie sessions, the shared state is also stored independently in the cookie (per request). So when you have multiple requests running in parallel, each has an independent view of the session state.

The double session instances are less of a problem as you could work around that by getting the session in the parent and passing it along as loader context I think.

That wat the child could get the session from the loader context instead of deriving it from the request.

> Another option is to simply return the error message from your loader
I'm already doing that for some other form validation errors, but your proposed solution for the `Toast` would only work if the `Toast` would have a centralised stack.

> P.S. Session flash is good if you need to pass a value/message across redirects. But if you're on the same route, it's typically better to just return that data in the loader response.

In that sense I've got something along the lines of the official [toast-message example](https://github.com/remix-run/remix/tree/main/examples/toast-message) where I read out the flash messages on a more global level.

Edit: SO reference for the people reading this later on: https://stackoverflow.com/questions/73204981/mutate-same-session-cookie-in-different-loaders

**Reply by kiliman** [reply]
> The double session instances are less of a problem as you could work around that by getting the session in the parent and passing it along as loader context I think.

I'm not sure what you mean by *loader context* here. Remix runs all loaders in parallel, so you can't pass data from parent to child, they're both running at the same time. If you're referring to `getLoadContext`, that is provided by your server entry point before handing the request to Remix. <https://remix.run/docs/en/v1/api/conventions#loader-context>

**Reply by kiliman** [reply]
Even though you could stick stuff on the context object from your loader, there's no guarantee that your loaders will all run in the same process: serverless for example, or load-balanced node instances. Although Remix has a lot of flexibility, I would caution against doing anything "cute" to work around the Remix model.

**Comment by davecranwell-vocovo** [comment]
I'm struggling with this too, weirdly at roughly the same time.

My use-case of this is refreshing an API token just prior to firing an API call. Literally any route, at any depth, which uses my API might need to set a new cookie having just got a new token which it wants to commit to the session.

But imagine this API call is run from a loader in a page **parallel** to another also trying to display a flash message from some previous action. Reading flash messages requires writing a cookie. The same cookie!

Naturally we've got ourselves a race-condition.

So far i've managed to keep the app i'm working on free of all dependencies (on the front end) besides Node to serve remix. I'm loath to have to install a database to manage sessions.

**Reply by kiliman** [reply]
I'm working on a file based session storage that supports locking, so only one writer can update session at a time to prevent corruption. Each commit tries to acquire the lock, and if lock already acquired, it will wait and try again.

**Reply by davecranwell-vocovo** [reply]
Ooh, looking forward to that!

**Reply by davecranwell-vocovo** [reply]
Any idea if the core remix file-session is getting a fix? It seems pretty obviously broken with the most elementary of use-cases. Will your solution be worked into core?

**Reply by kiliman** [reply]
Once it's been used for a while, I'll probably submit a PR.

**Reply by ToneVDB** [reply]
The file based session storage would not work on a distributed system like AWS lambda, am I right? Without plugging in on an other AWS service that is.

**Reply by kiliman** [reply]
No. You'd need a single file system instance. For serverless, I would use some sort of database, redis cache, etc.
